error: cannot use `#[pyo3(set)]` on a `frozen` class
  --> tests/ui/invalid_frozen_pyclass_borrow.rs:38:12
   |
38 |     #[pyo3(set)]
   |            ^^^

error[E0271]: type mismatch resolving `<Foo as PyClass>::Frozen == False`
  --> tests/ui/invalid_frozen_pyclass_borrow.rs:11:19
   |
11 |     fn mut_method(&mut self) {}
   |                   ^ type mismatch resolving `<Foo as PyClass>::Frozen == False`
   |
note: expected this to be `False`
  --> tests/ui/invalid_frozen_pyclass_borrow.rs:3:1
   |
3  | #[pyclass(frozen)]
   | ^^^^^^^^^^^^^^^^^^
note: required by a bound in `extract_pyclass_ref_mut`
  --> src/impl_/extract_argument.rs
   |
   | pub fn extract_pyclass_ref_mut<'a, 'holder, T: PyClass<Frozen = False>>(
   |                                                        ^^^^^^^^^^^^^^ required by this bound in `extract_pyclass_ref_mut`
   = note: this error originates in the attribute macro `pyclass` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0271]: type mismatch resolving `<Foo as PyClass>::Frozen == False`
 --> tests/ui/invalid_frozen_pyclass_borrow.rs:9:1
  |
9 | #[pymethods]
  | ^^^^^^^^^^^^ type mismatch resolving `<Foo as PyClass>::Frozen == False`
  |
note: expected this to be `False`
 --> tests/ui/invalid_frozen_pyclass_borrow.rs:3:1
  |
3 | #[pyclass(frozen)]
  | ^^^^^^^^^^^^^^^^^^
note: required by a bound in `PyClassGuardMut`
 --> src/pyclass/guard.rs
  |
  | pub struct PyClassGuardMut<'a, T: PyClass<Frozen = False>> {
  |                                           ^^^^^^^^^^^^^^ required by this bound in `PyClassGuardMut`
  = note: this error originates in the attribute macro `pymethods` which comes from the expansion of the attribute macro `pyclass` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0271]: type mismatch resolving `<Foo as PyClass>::Frozen == False`
  --> tests/ui/invalid_frozen_pyclass_borrow.rs:15:31
   |
15 |     let borrow = foo.bind(py).borrow_mut();
   |                               ^^^^^^^^^^ type mismatch resolving `<Foo as PyClass>::Frozen == False`
   |
note: expected this to be `False`
  --> tests/ui/invalid_frozen_pyclass_borrow.rs:3:1
   |
3  | #[pyclass(frozen)]
   | ^^^^^^^^^^^^^^^^^^
note: required by a bound in `pyo3::Bound::<'py, T>::borrow_mut`
  --> src/instance.rs
   |
   |     pub fn borrow_mut(&self) -> PyRefMut<'py, T>
   |            ---------- required by a bound in this associated function
   |     where
   |         T: PyClass<Frozen = False>,
   |                    ^^^^^^^^^^^^^^ required by this bound in `Bound::<'py, T>::borrow_mut`
   = note: this error originates in the attribute macro `pyclass` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0271]: type mismatch resolving `<ImmutableChild as PyClass>::Frozen == False`
  --> tests/ui/invalid_frozen_pyclass_borrow.rs:25:33
   |
25 |     let borrow = child.bind(py).borrow_mut();
   |                                 ^^^^^^^^^^ type mismatch resolving `<ImmutableChild as PyClass>::Frozen == False`
   |
note: expected this to be `False`
  --> tests/ui/invalid_frozen_pyclass_borrow.rs:21:1
   |
21 | #[pyclass(frozen, extends = MutableBase)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: required by a bound in `pyo3::Bound::<'py, T>::borrow_mut`
  --> src/instance.rs
   |
   |     pub fn borrow_mut(&self) -> PyRefMut<'py, T>
   |            ---------- required by a bound in this associated function
   |     where
   |         T: PyClass<Frozen = False>,
   |                    ^^^^^^^^^^^^^^ required by this bound in `Bound::<'py, T>::borrow_mut`
   = note: this error originates in the attribute macro `pyclass` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0271]: type mismatch resolving `<MutableBase as PyClass>::Frozen == True`
  --> tests/ui/invalid_frozen_pyclass_borrow.rs:29:11
   |
29 |     class.get();
   |           ^^^ type mismatch resolving `<MutableBase as PyClass>::Frozen == True`
   |
note: expected this to be `True`
  --> tests/ui/invalid_frozen_pyclass_borrow.rs:18:1
   |
18 | #[pyclass(subclass)]
   | ^^^^^^^^^^^^^^^^^^^^
note: required by a bound in `pyo3::Py::<T>::get`
  --> src/instance.rs
   |
   |     pub fn get(&self) -> &T
   |            --- required by a bound in this associated function
   |     where
   |         T: PyClass<Frozen = True> + Sync,
   |                    ^^^^^^^^^^^^^ required by this bound in `Py::<T>::get`
   = note: this error originates in the attribute macro `pyclass` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0271]: type mismatch resolving `<MutableBase as PyClass>::Frozen == True`
  --> tests/ui/invalid_frozen_pyclass_borrow.rs:33:11
   |
33 |     class.get();
   |           ^^^ type mismatch resolving `<MutableBase as PyClass>::Frozen == True`
   |
note: expected this to be `True`
  --> tests/ui/invalid_frozen_pyclass_borrow.rs:18:1
   |
18 | #[pyclass(subclass)]
   | ^^^^^^^^^^^^^^^^^^^^
note: required by a bound in `pyo3::Bound::<'py, T>::get`
  --> src/instance.rs
   |
   |     pub fn get(&self) -> &T
   |            --- required by a bound in this associated function
   |     where
   |         T: PyClass<Frozen = True> + Sync,
   |                    ^^^^^^^^^^^^^ required by this bound in `Bound::<'py, T>::get`
   = note: this error originates in the attribute macro `pyclass` (in Nightly builds, run with -Z macro-backtrace for more info)
